sim4rec/response/nn_response.py

```diff
@@ -0,0 +1,204 @@
+import os
+import pickle
+import pyspark.sql.functions as sf
+
+from .response import ActionModelEstimator, ActionModelTransformer
+from .nn_utils.models import ResponseModel
+from .nn_utils.embeddings import IndexEmbedding
+from .nn_utils.datasets import (
+    RecommendationData,
+    # PandasRecommendationData,
+)
+
+from pyspark.sql.types import (
+    StructType,
+    StructField,
+    IntegerType,
+    DoubleType,
+)
+
+# move this to simulator core(?)
+SIM_LOG_SCHEMA = StructType(
+    [
+        StructField("user_idx", IntegerType(), True),
+        StructField("item_idx", IntegerType(), True),
+        StructField("relevance", DoubleType(), True),
+        StructField("response_proba", DoubleType(), True),
+        StructField("response", IntegerType(), True),
+        StructField("__iter", IntegerType(), True),
+    ]
+)
+SIM_LOG_COLS = [field.name for field in SIM_LOG_SCHEMA.fields]
+
+
+class NNResponseTransformer(ActionModelTransformer):
+    def __init__(self, **kwargs):
+        super().__init__()
+        self.hist_data = None
+        for param, value in kwargs.items():
+            setattr(self, param, value)
+
+    @classmethod
+    def load(cls, checkpoint_dir):
+        with open(os.path.join(checkpoint_dir, "_params.pkl"), "rb") as f:
+            params_dict = pickle.load(f)
+        params_dict["backbone_response_model"] = ResponseModel.load(checkpoint_dir)
+        with open(os.path.join(checkpoint_dir, "_item_indexer.pkl"), "rb") as f:
+            params_dict["item_indexer"] = pickle.load(f)
+        with open(os.path.join(checkpoint_dir, "_user_indexer.pkl"), "rb") as f:
+            params_dict["user_indexer"] = pickle.load(f)
+        return cls(**params_dict)
+
+    def save(self, path):
+        """Save model at given path."""
+        os.makedirs(path)
+        self.backbone_response_model.dump(path)
+        with open(os.path.join(path, "_item_indexer.pkl"), "wb") as f:
+            pickle.dump(self.item_indexer, f, pickle.HIGHEST_PROTOCOL)
+        with open(os.path.join(path, "_user_indexer.pkl"), "wb") as f:
+            pickle.dump(self.user_indexer, f, pickle.HIGHEST_PROTOCOL)
+        with open(os.path.join(path, "_params.pkl"), "wb") as f:
+            pickle.dump(
+                {
+                    "outputCol": self.outputCol,
+                    "log_dir": self.log_dir,
+                    "hist_data_dir": self.hist_data_dir,
+                },
+                f,
+                pickle.HIGHEST_PROTOCOL,
+            )
+
+    def _transform(self, new_recs):
+        """
+        Predict responses for given dataframe with recommendations.
+
+        :param dataframe: new recommendations.
+        """
+
+        def predict_udf(df):
+            # if not do this, something unstable happens to the Method Resolution Order
```

Comment:
what exactly?

Reply:
I don't really understans why, seems it is related on how pySpark sends serialized python objects to workers. The effect is following: my datasets looses inheritance from DatasetBase, and become just exempplars of PandasRecommendationData(object) instead. And when the base class method is called, everything falls with an error.